with the branch for this PR, I am getting an additional error
```
Execution platform: @local_execution_config_platform//:platform
tensorflow/compiler/mlir/tools/kernel_gen/transforms/gpu_kernel_to_blob_pass.cc: In member function 'xla::StatusOr<std::vector<unsigned char, std::allocator<unsigned char> > > mlir::kernel_gen::transforms::{anonymous}::GpuKernelToBlobPass::GetGpuBinaryBlob(mlir::gpu::GPUModuleOp)':
tensorflow/compiler/mlir/tools/kernel_gen/transforms/gpu_kernel_to_blob_pass.cc:124:62: error: could not convert 'arch' from 'uint32_t {aka unsigned int}' to 'xla::gpu::GpuVersion {aka absl::lts_2020_02_25::variant<std::pair<int, int>, std::pair<int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >}'
           llvm_module_copy.get(), arch, config, libdevice_dir);
                                                              ^
tensorflow/compiler/mlir/tools/kernel_gen/transforms/gpu_kernel_to_blob_pass.cc:137:52: error: no matching function for call to 'std::vector<stream_executor::HsacoImage>::push_back(<brace-enclosed initializer list>)'
       images.push_back({arch_str, std::move(hsaco)});
                                                    ^
I
```

working on fixing it now

--- a/tensorflow/compiler/mlir/tools/kernel_gen/transforms/gpu_kernel_to_blob_pass.cc
+++ b/tensorflow/compiler/mlir/tools/kernel_gen/transforms/gpu_kernel_to_blob_pass.cc
@@ -120,8 +120,9 @@ class GpuKernelToBlobPass
 
       std::string libdevice_dir = tensorflow::RocdlRoot();
       auto llvm_module_copy = llvm::CloneModule(*llvmModule);
+      xla::gpu::GpuVersion gpu_version{std::make_pair(arch, arch_str)};
       auto hsaco_or = xla::gpu::amdgpu::CompileToHsaco(
-          llvm_module_copy.get(), arch, config, libdevice_dir);
+          llvm_module_copy.get(), gpu_version, config, libdevice_dir);
       if (!hsaco_or.ok()) {
         return InternalError("Failure when generating HSACO");
       }

--- a/tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc
+++ b/tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc
@@ -809,14 +809,15 @@ Status AMDGPUTargetModuleLinker(llvm::Module* module, GpuVersion gpu_version,
 // When that upstreaming happens (and TF LLVM pointer moves past the
 // upstream commit), the following mapping will need to change
 static std::string MapGCNArchNameTokenToFeatureStr(const std::string token) {
-  if (token == "sramecc+")
+  if (token == "sramecc+") {
     return "+sram-ecc";
-  else if (token == "sramecc-")
+  } else if (token == "sramecc-") {
     return "-sram-ecc";
-  else if (token == "xnack+")
+  } else if (token == "xnack+") {
     return "+xnack";
-  else if (token == "xnack-")
+  } else if (token == "xnack-") {
     return "-xnack";
+  }
   return "";
 }
 
@@ -833,23 +834,19 @@ static std::string GetFeatureStrFromGCNArchName(
   feature_str = "";
 #else
   // For ROCm versions 4.0 and greater, we need to specify the correct
-  // feature str, based on the underlying GPU WH to get max performance.
-  feature_str = "";
-  int token_num = 0;
-  size_t start = 0, pos = 0;
-  do {
-    pos = gcn_arch_name.find(":", start);
-    if (token_num != 0) {
-      // first token will be "gfxNNN"...ignore that, use the rest
-      if (feature_str.size()) feature_str += ",";
-      std::string token = gcn_arch_name.substr(start, (pos == std::string::npos)
-                                                          ? std::string::npos
-                                                          : (pos - start));
-      feature_str += MapGCNArchNameTokenToFeatureStr(token);
+  // feature str, based on the underlying GPU HW to get max performance.
+  std::vector<std::string> tokens = absl::StrSplit(gcn_arch_name, ':');
+  std::vector<std::string> mapped_tokens;
+  for (auto it = tokens.begin(); it != tokens.end(); it++) {
+    // Skip the first token, that is the gfxNNN str
+    // The rest of the tokens are the feature/targetid strings
+    if (it != tokens.begin()) {
+      std::string token(*it);
+      std::string mapped_token = MapGCNArchNameTokenToFeatureStr(token);
+      mapped_tokens.push_back(mapped_token);
     }
-    start = pos + 1;
-    token_num++;
-  } while (pos != std::string::npos);
+  }
+  feature_str = absl::StrJoin(mapped_tokens, ",");
 #endif
 
   return feature_str;

